### PR TITLE
Add a CSV reader that doesn't need to know number of rows or columns of strings

### PR DESCRIPTION
**Pull request by daviditen**
Add a CSV reader that doesn't need to know the number of rows or columns in
advance, but only works with string data.

### CODE REVIEW

**Comment by daviditen** [comment]
Passed full Linux64 testing.

File: test/library/packages/csv/CSV.chpl [hunk 1]
```diff
@@ -69,7 +69,10 @@ module CSV {
     /* Read a CSV file with fields of the types given by
        the arguments to the function
      */
-    iter read(type t...) throws {
+    iter read(type t...) throws where t.size > 1 ||
```
Comment (ben-albrecht) [inline test/library/packages/csv/CSV.chpl]: Would it make sense to wrap the where clause into a helper function? Maybe this list of cases to check will grow as we support more types?  I could be wrong.

File: test/library/packages/csv/CSV.chpl [hunk 2]
```diff
@@ -90,11 +93,44 @@ module CSV {
...
+      var r: t;
+      var skipHeader = hasHeader;
+
+      var lines = ch.lines();
+      var firstLine = lines[1];
+      var vals = firstLine.strip().split(sep);
+      var A: [1..lines.numElements, 1..vals.numElements] string;
+
+      if !skipHeader {
+        A[1, ..] = vals;
```
Comment (ben-albrecht) [inline test/library/packages/csv/CSV.chpl]: When this is skipped, does the returned array have an initial row of 0s?

Comment (ben-albrecht) [reply test/library/packages/csv/CSV.chpl]: Could be worth adding a test for this case too.

File: test/library/packages/csv/readStrings.chpl [hunk 3]
```diff
@@ -0,0 +1,11 @@
+use CSV;
+
+config const filename = "strings.csv";
+
+var reader = openreader(filename);
+var r = new CSVIO(reader, sep=" ");
```
Comment (ben-albrecht) [inline test/library/packages/csv/readStrings.chpl]: This will be very useful!

File: test/library/packages/csv/CSV.chpl (hunk 1 above)
Comment (daviditen) [reply test/library/packages/csv/CSV.chpl]: I improved the where clauses.

File: test/library/packages/csv/CSV.chpl (hunk 2 above)
Comment (daviditen) [reply test/library/packages/csv/CSV.chpl]: It did have a row of empty strings.  I fixed it to be one row smaller and added a test.

**Comment by daviditen** [comment]
Thanks!

**Pull request merged**